src/build/compile/dependency_cycle.rs

```diff
@@ -1,71 +1,152 @@
 use super::super::build_types::*;
 use crate::helpers;
-use ahash::AHashSet;
+use std::collections::{HashMap, HashSet, VecDeque};
 
 pub fn find(modules: &Vec<(&String, &Module)>) -> Vec<String> {
-    let mut visited: AHashSet<String> = AHashSet::new();
-    let mut stack: Vec<String> = vec![];
+    find_shortest_cycle(modules)
+}
 
-    // we want to sort the module names so that we always return the same
-    // dependency cycle (there can be more than one)
-    let mut module_names = modules
-        .iter()
-        .map(|(name, _)| name.to_string())
-        .collect::<Vec<String>>();
+fn find_shortest_cycle(modules: &Vec<(&String, &Module)>) -> Vec<String> {
+    let mut shortest_cycle: Vec<String> = Vec::new();
+
+    // Build a graph representation for easier traversal
+    let mut graph: HashMap<String, Vec<String>> = HashMap::new();
+    let mut in_degrees: HashMap<String, usize> = HashMap::new();
+
+    // First pass: collect all nodes and initialize in-degrees
+    for (name, _) in modules {
+        graph.insert(name.to_string(), Vec::new());
+        in_degrees.insert(name.to_string(), 0);
+    }
 
-    module_names.sort();
-    for module_name in module_names {
-        if find_dependency_cycle_helper(&module_name, modules, &mut visited, &mut stack) {
-            return stack;
+    // Second pass: build the graph and count in-degrees
+    for (name, module) in modules {
+        let deps: Vec<String> = module.deps.iter().cloned().collect();
+
+        // Update the graph
+        *graph.get_mut(&name.to_string()).unwrap() = deps.clone();
+
+        // Update in-degrees
+        for dep in deps {
+            if let Some(count) = in_degrees.get_mut(&dep) {
+                *count += 1;
+            }
         }
-        visited.clear();
-        stack.clear();
     }
-    stack
-}
 
-fn find_dependency_cycle_helper(
-    module_name: &String,
-    modules: &Vec<(&String, &Module)>,
-    visited: &mut AHashSet<String>,
-    stack: &mut Vec<String>,
-) -> bool {
-    if let Some(module) = modules
-        .iter()
-        .find(|(name, _)| *name == module_name)
-        .map(|(_, module)| module)
-    {
-        visited.insert(module_name.to_string());
-        // if the module is a mlmap (namespace), we don't want to show this in the path
-        // because the namespace is not a module the user created, so only add source files
-        // to the stack
-        if let SourceType::SourceFile(_) = module.source_type {
-            stack.push(module_name.to_string())
+    // OPTIMIZATION 1: Start with nodes that are more likely to be in cycles
+    // Sort nodes by their connectivity (in-degree + out-degree)
+    let mut start_nodes: Vec<String> = graph.keys().cloned().collect();
```

Comment:
I think you can first filter out all graph keys that don't have any edges connected to them here. I think the check might be happening later, but we can just drop them here.

Also - I'm not sure if you need to clone these here